### ISSUE TITLE
[models] allow specifying block numbers in `enable_gradient_checkpointing()` to speed up training

### ISSUE THREAD

**Issue by sayakpaul** [huggingface/diffusers#10124]
As brought up in https://github.com/huggingface/diffusers/pull/9982#issuecomment-2517826764 by @bghira, I think we could support this directly in the `enable_gradient_checkpointing()` method we expose for the models. 

Users could specify the block interval they want gradient checkpointing to be applied in and we take care of the rest. The code for this is simple and doesn't require any hacks. 

Gradient checkpointing is a crucial component for training/fine-tuning larger models and this technique allows for nice speed/memory trade-off. 

Cc: @a-r-r-o-w @hlky

**Comment by a-r-r-o-w** [comment]
Taking this up soon. Will prio for next release

**Comment by bghira** [comment]
i've implemented this in simpletuner's modeling code for SDXL, Flux, SD3, and Sana, and even the smaller models benefit (maybe moreso) from having this option. this means that basically Sana goes from 4 seconds per step to 1.25 steps per second on a 24G card (int8) when training. I actually haven't checked the speedup for Flux but I assume on 80G card it'll be quite substantial.
